app/models/build.rb

```diff
@@ -79,8 +79,10 @@ def validate_git_reference
 
     return if errors.include?(:git_ref) || errors.include?(:git_sha)
 
-    project.with_lock(holder: 'Build reference validation') do
-      project.repository.setup_local_cache!
+    unless project.repository.last_pulled
+      project.with_lock(holder: 'Build reference validation') do
+        project.repository.setup_local_cache!
+      end
     end
```

Comment:
Without this check, it would end up doing two `git fetch` operations each time a Build was created, and that's unnecessary overhead.